### DISCUSSION TITLE
How to retrieve the actual JWT secret and valid JWT tokens?

### DISCUSSION THREAD

**Discussion by gusfune** [payloadcms/payload#2575] [Q&A]
Hi,
I am trying to connect PayloadCMS with a separate system that decodes JWT and can also use that for RBAC, but, I am having problem with my generated JWT after login.

After I login an user programmatically, the generated JWT is valid for PayloadCMS, but invalid for external systems. For example, if I test it at https://jwt.io/ is shows as invalid.

Also, to connect to an external system, I'd need the actual secret key generated/hashed, [as responded here](https://github.com/payloadcms/payload/discussions/347) to seamless integrate the system and decode the JWT.

Any suggestions on how to do that?

Here is an example JWT that is generated by going to `http://localhost:3000/api/users/login`:
```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImhlbGxvQHdvcmxkLmNvbSIsImlkIjoiNjQ0YWI4YjU5MGZmODFhNGNiOWJkMTlmIiwiY29sbGVjdGlvbiI6InVzZXJzIiwiaWF0IjoxNjgyNjE4NTU3LCJleHAiOjE2ODI2MjU3NTd9.BKGq3niekGKnwuItyK4JmIcF4vKaK5ARstAw10F1zcQ
```

https://jwt.io/#debugger-io?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImhlbGxvQHdvcmxkLmNvbSIsImlkIjoiNjQ0YWI4YjU5MGZmODFhNGNiOWJkMTlmIiwiY29sbGVjdGlvbiI6InVzZXJzIiwiaWF0IjoxNjgyNjE4NTU3LCJleHAiOjE2ODI2MjU3NTd9.BKGq3niekGKnwuItyK4JmIcF4vKaK5ARstAw10F1zcQ

**Comment by dunganh21** [comment]
PayloadCMS hashes the JWT secret using SHA-256, truncates it to 32 characters, and uses that for signing. To decode the JWT externally, you need to hash your secret the same way. Here's how
```
import { jwtVerify } from "jose";
import crypto from "crypto";

const testDecodeJwt = async (token: string, secret: string) => {
  const newSecret = crypto
    .createHash("sha256")
    .update(secret)
    .digest("hex")
    .slice(0, 32);
  const secretKey = new TextEncoder().encode(newSecret);
  const { payload: decodedPayload } = await jwtVerify(token, secretKey);
  console.log(decodedPayload);
};
```
In case you still need it

**Reply by didiraja** [reply]
The function above is very very helpful. Unfortunately, doesn't work on Edge runtime that doesn't have "crypto" api, when trying to run on Next middleware file for example. Here is an example to make it run on middlewares:

```typescript
export const decodePayloadToken = async (token: string): Promise<any | null> => {
  const secret = process.env.PAYLOAD_SECRET
  if (!secret || secret.length === 0) return null

  try {
    const encoder = new TextEncoder()
    const secretEncoded = encoder.encode(secret)
    const hashBuffer = await crypto.subtle.digest('SHA-256', secretEncoded)

    const hashHex = Array.from(new Uint8Array(hashBuffer))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('')

    const shortHash = hashHex.slice(0, 32)
    const key = encoder.encode(shortHash)

    const { payload } = await jwtVerify(token, key)
    return payload
  } catch (e) {
    console.error('Error decoding token:', e)
    return null
  }
}
```
(Sometimes jwtVerify returns `code: 'ERR_JWS_INVALID'` but it passes anyway. Couldn't fix this.)


Bonus: Since we're on the topic of Payload tokens, here's an article that breaks down how they work and how to build your own custom tokens. [Custom JWT Token Generation in Payload CMS: A Step-by-Step Guide](https://medium.com/@nirjalprajapati/custom-jwt-token-generation-in-payload-cms-a-step-by-step-guide-ccb1546fa8bf)
